core-macros/src/main/scala/sbt/internal/util/appmacro/Cont.scala

```diff
@@ -203,57 +266,140 @@ trait Cont:
-              val substitute = [x] =>
-                (name: String, tpe: Type[x], qual: Term, oldTree: Term) =>
-                  given Type[x] = tpe
-                  convert[x](name, qual) transform { (replacement: Term) =>
-                    val idx = inputs.indexWhere(input => input.qual == qual)
-                    Select
-                      .unique(Ref(p0.symbol), "apply")
-                      .appliedToTypes(List(br.inputTupleTypeRepr))
-                      .appliedToArgs(List(Literal(IntConstant(idx))))
+          br.inputTupleTypeRepr.asType match
+            case '[inputTypeTpe] =>
+              val lambda = Lambda(
+                owner = Symbol.spliceOwner,
+                tpe = lambdaTpe,
+                rhsFn = (sym, params) => {
+                  val p0 = params.head.asInstanceOf[Term]
+                  // Called when transforming the tree to add an input.
+                  //  For `qual` of type F[A], and a `selection` qual.value,
+                  //  the call is addType(Type A, Tree qual)
+                  // The result is a Tree representing a reference to
+                  //  the bound value of the input.
+                  val substitute = [x] =>
+                    (name: String, tpe: Type[x], qual: Term, oldTree: Term) =>
+                      given Type[x] = tpe
+                      convert[x](name, qual) transform { (replacement: Term) =>
+                        val idx = inputs.indexWhere(input => input.qual == qual)
+                        applyTuple(p0, br.inputTupleTypeRepr, idx)
+                    }
+                  val modifiedBody =
+                    transformWrappers(body.asTerm.changeOwner(sym), substitute, sym).asExprOf[A1]
+                  cacheConfigExprOpt match
+                    case Some(cacheConfigExpr) =>
+                      if inputs.exists(_.isCacheInput) then
+                        val tags = inputs
+                          .filter(_.isCacheInput)
+                          .map(_.tags.toSet)
+                          .reduce(_ & _)
+                          .toList
+                        require(
+                          tags.nonEmpty,
+                          s"""cacheLevelTag union must be non-empty: ${inputs.mkString("\n")}"""
+                        )
+                        br.cacheInputTupleTypeRepr.asType match
+                          case '[cacheInputTpe] =>
+                            callActionCacheWithA2(outputBuf.toList, cacheConfigExpr, tags)(
+                              body = modifiedBody,
+                              input = br.cacheInputExpr(p0).asExprOf[cacheInputTpe],
+                            ).asTerm.changeOwner(sym)
+                      else
+                        val tags = CacheLevelTag.all.toList
+                        callActionCacheWithUnit(outputBuf.toList, cacheConfigExpr, tags)(
+                          modifiedBody
+                        ).asTerm.changeOwner(sym)
```

Comment:
We don't need to have `callActionCacheWithA2` and `callActionCacheWithUnit`, just to save one argument once. 

```suggestion
                            callActionCache(outputBuf.toList, cacheConfigExpr, tags)(
                              body = modifiedBody,
                              input = br.cacheInputExpr(p0).asExprOf[cacheInputTpe],
                            ).asTerm.changeOwner(sym)
                      else
                        val tags = CacheLevelTag.all.toList
                        callActionCache(outputBuf.toList, cacheConfigExpr, tags)(
                          modifiedBody,
                          '{ ( ) },
                        ).asTerm.changeOwner(sym)
```

Reply:
I fixed it in c704fe357e07aae37f14caedbea74a08387cdeb0, but note `callActionCacheWithUnit` appears in the code three times, and usually threepeat is a signal to refactor.